packages/picasso.js/src/core/chart-components/box/box-shapes.js

```diff
@@ -208,114 +209,174 @@ export function buildShapes({
   // }
 
   const output = [];
+  const majorItems = resolved.major.items;
 
-  let major = null;
-  const items = resolved.major.items;
-
-  for (let i = 0, len = items.length; i < len; i++) {
-    const d = items[i].data;
-    let children = [];
-
-    let majorVal = null;
-    let majorEndVal = null;
-
-    if (typeof resolved.major.settings.binStart !== 'undefined') { // if start and end is defined
-      majorVal = resolved.major.items[i].binStart;
-      majorEndVal = resolved.major.items[i].binEnd;
-      major = resolved.major.settings.binStart.scale;
-    } else {
-      major = resolved.major.settings.major.scale;
-      majorVal = major ? resolved.major.items[i].major : 0;
-    }
-
-    let bandwidth = 0;
-    if (!major) {
-      bandwidth = 1;
-    } else if (major.bandwidth) {
-      bandwidth = major.bandwidth();
-      majorVal -= (bandwidth / 2);
-    } else {
-      bandwidth = majorEndVal - majorVal;
-    }
-
-    let item = extend({}, {
-      major: majorVal,
-      majorEnd: majorEndVal
-    }, resolved.minor.items[i]);
-
-    keys.forEach(key => (item[key] = resolved[key].items[i]));
-
-    const maxMajorWidth = flipXY ? height : width;
-    const boxWidth = getBoxWidth(bandwidth, item, maxMajorWidth);
-    const boxPadding = (bandwidth - boxWidth) / 2;
-    const boxCenter = boxPadding + item.major + (boxWidth / 2);
-
-    const rendWidth = width;
-    const rendHeight = height;
-
-
-    const allValidValues = [item.min, item.start, item.med, item.end, item.max].filter(isNumber);
-
-    const isLowerOutOfBounds = Math.min(...allValidValues) < 0 && Math.max(...allValidValues) < 0;
-    const isHigherOutOfBounds = Math.min(...allValidValues) > 1 && Math.max(...allValidValues) > 1;
-    const isOutOfBounds = isLowerOutOfBounds || isHigherOutOfBounds;
-
-    /* OUT OF BOUNDS */
-    if (item.oob.show && isLowerOutOfBounds) {
-      children.push(oob({
-        item, value: 0, boxCenter, rendWidth, rendHeight, flipXY, symbol
-      }));
-    }
-
-    if (item.oob.show && isHigherOutOfBounds) {
-      children.push(oob({
-        item, value: 1, boxCenter, rendWidth, rendHeight, flipXY, symbol
-      }));
-    }
+  if (!majorItems.length) {
+    return output;
+  }
 
+  const rendWidth = width;
+  const rendHeight = height;
+  const maxMajorWidth = flipXY ? height : width;
+  const majorSettings = resolved.major.settings;
+  const minorProps = ['start', 'end', 'min', 'max', 'med'].filter(prop => typeof resolved.minor.settings[prop] !== 'undefined');
+  const numMinorProps = minorProps.length;
+  const nonOobKeys = keys.filter(key => key !== 'oob');
+
+  let children;
+  let major;
+  let minorItem;
+  let boxWidth;
+  let boxPadding;
+  let boxCenter;
+  let isLowerOutOfBounds;
+  let isHigherOutOfBounds;
+  let isOutOfBounds;
+  const numKeys = keys ? keys.length : 0;
+  const numNonOobKeys = nonOobKeys ? nonOobKeys.length : 0;
+
+  function addBox() {
```

Comment:
Why not go all the way and break out these "add" functions from buildShapes? As buildShapes is now a ~200 lines function with a lot of complex logic that should be tested separately.